pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go

```diff
@@ -0,0 +1,156 @@
+	"golang.org/x/net/context"
+
+	registerapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1alpha1"
+	v1beta1 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta1"
+	v1beta2 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta2"
+)
+
+func TestExamplePlugin(t *testing.T) {
+	socketDir, err := ioutil.TempDir("", "plugin_test")
+	require.NoError(t, err)
+	socketPath := socketDir + "/plugin.sock"
+	w := NewWatcher(socketDir)
+
+	testCases := []struct {
+		description string
+		returnErr   error
+	}{
+		{
+			description: "Successfully register plugin through inotify",
+			returnErr:   nil,
+		},
+		{
+			description: "Successfully register plugin through inotify after plugin restarts",
+			returnErr:   nil,
+		},
+		{
+			description: "Fails registration with conflicting plugin name",
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+		{
+			description: "Successfully register plugin during initial traverse after plugin watcher restarts",
+			returnErr:   nil,
+		},
+		{
+			description: "Fails registration with conflicting plugin name during initial traverse after plugin watcher restarts",
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+	}
+
+	callbackCount := 0
+	w.AddHandler(PluginType, func(name string, versions []string, sockPath string) error {
+		require.True(t, callbackCount <= (len(testCases)-1))
+		glog.Infof("receives plugin watcher callback for test: %s\n", testCases[callbackCount].description)
+		require.Equal(t, PluginName, name, "Plugin name mismatched!!")
+		require.Equal(t, []string{"v1beta1", "v1beta2"}, versions, "Plugin version mismatched!!")
+		// Verifies the grpcServer is ready to serve services.
+		_, conn, err := dial(socketPath)
+		require.Nil(t, err)
+		defer conn.Close()
+
+		// The plugin handler should be able to use any listed service API version.
+		v1beta1Client := v1beta1.NewExampleClient(conn)
+		v1beta2Client := v1beta2.NewExampleClient(conn)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta1Client.GetExampleInfo(context.Background(), &v1beta1.ExampleRequest{})
+		require.Nil(t, err)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta2Client.GetExampleInfo(context.Background(), &v1beta2.ExampleRequest{})
+		ret := testCases[callbackCount].returnErr
+		callbackCount++
+		//require.True(t, true)
+		return ret
+	})
+	err = w.Start()
+	require.NoError(t, err)
+
+	p := NewTestExamplePlugin() //first cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+
+	// Trying to start a plugin service at the same socket path should fail
+	// with "bind: address already in use"
+	err = p.Serve(socketPath)
+	require.NotNil(t, err)
+
+	// grpcServer.Stop() will remove the socket and starting plugin service
+	// at the same path again should succeeds and trigger another callback.
+	p.Stop()
+	p = NewTestExamplePlugin() // 2nd cbk
+	go func() {
+		p.Serve(socketPath)
```

Comment:
Looking at how serve is used it probably should start it's own goroutine and not be blocking